readme.md
[![Build Status](https://app.travis-ci.com/AndreasKl/reactive-spring-session-vertx.svg?branch=master)](https://app.travis-ci.com/AndreasKl/reactive-spring-session-vertx)
[![Maven Central](https://img.shields.io/maven-central/v/net.andreaskluth/reactive-spring-session-postgres.svg?label=Maven%20Central)](https://search.maven.org/search?q=g:%22net.andreaskluth%22%20AND%20a:%22reactive-spring-session-postgres%22)
[![javadoc](https://javadoc.io/badge2/net.andreaskluth/reactive-spring-session-core/javadoc.svg)](https://javadoc.io/doc/net.andreaskluth/reactive-spring-session-core) 
# reactive-spring-session-vertx

Due to the lack of a distributed reactive Spring Session implementations for relational databases,
this projects aims to fill the gap with a fully reactive **PostgreSQL** any **MySQL** Spring Session store.

The implementation is based on **[vertx-sql-client](https://github.com/eclipse-vertx/vertx-sql-client)**.

Supports JDK 8 and up. Does not depend on **Spring Boot** but can be easily used with **Spring Boot**.

### A few words on MySQL
The MySQL flavour is compatible with MySQL 5.7 and up. It is fully working but not battle tested on production like the PostgreSQL version.

## Setup instructions
A demo project is located under `/reactive-spring-session-sample-app` written in Kotlin.
`ReactivePostgresSessionConfiguration` registers a `ReactivePostgresSessionRepository` 
and a scheduled task that removes expired sessions from the database. Expired sessions
are never returned to the user, however could remain in the database until the scheduler
does the cleanup.

```kotlin
@Configuration
@Import(ReactivePostgresSessionConfiguration::class)
class PostgresSessionConfiguration {

  @Bean
  fun pgConnectOptions(@Value("\${postgres.port}") postgresPort: Int): PgConnectOptions =
    PgConnectOptions()
      .setHost("localhost")
      .setPort(postgresPort)
      .setDatabase("session")
      .setUser("postgres")
      .setPassword("postgres")
      .setIdleTimeout(300)
      .setConnectTimeout(500)

  @Bean
  fun poolOptions(): PoolOptions =
    PoolOptions()
      .setMaxSize(5)
      .setMaxWaitQueueSize(10)

  @Bean
  fun clock(): Clock =
    Clock.systemUTC()

  @Bean(WebHttpHandlerBuilder.WEB_SESSION_MANAGER_BEAN_NAME)
  fun webSessionManager(repository: ReactiveSessionRepository<out Session>): WebSessionManager {
    val sessionStore = SpringSessionWebSessionStore(repository)

    val manager = DefaultWebSessionManager()
    manager.sessionStore = sessionStore
    return manager
  }

}
```

## Is this production ready?
**reactive-spring-session-vertx** is used in production **[@REWE digital](https://www.rewe-digital.com/)** for their Spring Gateway based reactive API gateway.

## Build

Just test and build, will start Postgres and MySQL as embedded databases:
```bash
mvn clean verify
```

Deploy the current SNAPSHOT to **oss.sonatype.org**:
```bash
mvn -P release -pl reactive-spring-session-postgres -am clean deploy
```

Deploy a RELEASE to **oss.sonatype.org**:
```bash
mvn release:prepare

mvn -P release clean deploy

mvn -P release nexus-staging:rc-list

mvn -P release -DstagingRepositoryId=netandreaskluth-<some-id>  nexus-staging:release
```

## Contributing
Source code formatting is checked with [fmt-maven-plugin](https://github.com/coveooss/fmt-maven-plugin). Configuration files for the formatter of your preferred IDE can be found [here](https://github.com/google/styleguide) a plugin for IntelliJ IDEA can be found [here](https://plugins.jetbrains.com/plugin/8527-google-java-format).

# License 
The MIT License (https://opensource.org/licenses/MIT)


changelog.md
# reactive-spring-session-vertx changelog

# 0.0.13
* Dependency updates:
  * VertX SQL 4.1.5
  * Spring Core 5.3.12
  * Spring Session 2.5.3
* Configured Enforcer Plugin

# 0.0.11
* Configure whether a deserialization exception invalidates the session or
  propagates the exception up to the caller.
  * See: net.andreaskluth.session.core.ReactiveVertxSessionRepository.setInvalidateSessionOnDeserializationError
* Support for Java8  
* Various dependency updates:
  * vertx-sql-client : 4.0.0-milestone5

# 0.0.10

# 0.0.9
* Removed jackson serializer, as this was always of limited value
* MySQL support
* Updated Spring Session 2.2.0.RELEASE and Spring Framework 5.2.0.RELEASE

# 0.0.8
* Update to vertx-sql-client 4.0.0-milestone3 to mitigate a direct buffer leak
* Update jackson-databind

# 0.0.7
Fixed metric name and tags.

# 0.0.6
* `ReactivePostgresSessionRepository` publishes metrics if enable. Set `ReactivePostgresSessionRepository#withMetrics` to `true` to activate `Mono#metrics`.
  * The `flow` tag can be configured by setting `ReactivePostgresSessionRepository#setSequenceName`.

# 0.0.1 - 0.0.5
* Initial releases used internally @rewe-digital



webflux-demo-application/src/main/kotlin/net/andreaskluth/reactivesession/PostgresSessionDemoApplication.kt
package net.andreaskluth.reactivesession

import com.opentable.db.postgres.embedded.ConnectionInfo
import com.opentable.db.postgres.embedded.PreparedDbProvider
import io.vertx.pgclient.PgConnectOptions
import io.vertx.sqlclient.Pool
import io.vertx.sqlclient.PoolOptions
import net.andreaskluth.session.core.support.ReactiveSessionSchemaPopulator
import net.andreaskluth.session.postgres.ReactivePostgresSessionConfiguration
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import
import org.springframework.http.ResponseEntity
import org.springframework.session.ReactiveSessionRepository
import org.springframework.session.Session
import org.springframework.session.web.server.session.SpringSessionWebSessionStore
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.server.adapter.WebHttpHandlerBuilder
import org.springframework.web.server.session.DefaultWebSessionManager
import org.springframework.web.server.session.WebSessionManager
import reactor.core.publisher.Mono
import java.time.Clock
import java.util.function.Consumer
import javax.annotation.PostConstruct

@SpringBootApplication
@Import(ReactivePostgresSessionConfiguration::class)
class PostgresSessionDemoApplication

@Configuration
class PostgresSessionConfiguration {

    @Bean
    fun pgConnectOptions(@Value("\${postgres.port}") postgresPort: Int): PgConnectOptions =
        PgConnectOptions()
            .setHost("localhost")
            .setPort(postgresPort)
            .setDatabase("session")
            .setUser("postgres")
            .setPassword("postgres")
            .setIdleTimeout(300)
            .setConnectTimeout(500)

    @Bean
    fun poolOptions(): PoolOptions =
        PoolOptions()
            .setMaxSize(5)
            .setMaxWaitQueueSize(10)

    @Bean
    fun clock(): Clock =
        Clock.systemUTC()

    @Bean(WebHttpHandlerBuilder.WEB_SESSION_MANAGER_BEAN_NAME)
    fun webSessionManager(repository: ReactiveSessionRepository<out Session>): WebSessionManager {
        val sessionStore = SpringSessionWebSessionStore(repository)

        val manager = DefaultWebSessionManager()
        manager.sessionStore = sessionStore
        return manager
    }

}

@Configuration
class PrepareSchemaConfiguration(val pool: Pool) {

    @PostConstruct
    fun prepareSchema() {
        ReactiveSessionSchemaPopulator.applyDefaultSchema(pool).block()
    }
}

@Controller
class HelloController {

    @GetMapping("/")
    fun hello(): ResponseEntity<Mono<String>> = ResponseEntity.ok(Mono.just("hallo"))
}

fun main(args: Array<String>) {
    val log = LoggerFactory.getLogger(PostgresSessionDemoApplication::class.java)

    val provider = PreparedDbProvider.forPreparer({ ds ->
        ds.connection.use { connection ->
            connection.createStatement().use { statement ->
                statement.execute("CREATE DATABASE session")
            }
        }
    }, listOf(Consumer { builder -> builder.setPort(39889) }))

    val connInfo = provider.createNewDatabase()
    exportPort(log, connInfo)
    runApplication<PostgresSessionDemoApplication>(*args)
}

private fun exportPort(log: Logger, connInfo: ConnectionInfo) {
    System.setProperty("postgres.port", connInfo.port.toString())
    log.info("Running with embedded postgres on port: {}", connInfo.port)
}
